Add geospatial averaging

tomvothecoder opened the pull request:
## Description
<!--
  Please include a summary of the change and which issue is fixed.
  Please also include relevant motivation and context.
  List any dependencies that are required for this change.
-->
This PR is a refactoring effort of PR #87, `feature/46-spatial-averaging`

- Closes #46 

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

If applicable:

- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass with my changes (locally and CI/CD build)
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -20,275 +23,311 @@
+SupportedAxes = Literal["lat", "lon"]
+SUPPORTED_AXES: Tuple[SupportedAxes, ...] = get_args(SupportedAxes)
 
-def _validate_region_lon_bounds(grid_lon_bounds, region_lon_bounds):
```
I broke this function up into the following:

1. `_validate_region_lon_bounds()`
2. `_swap_lon_region_axis()` (originally `_switch_lon_bounds()`
2. `_is_region_lon_in_domain()`

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -20,275 +23,311 @@
 # Should remove these notes and outline...
 
+#: List of floats representing user-specified region axis bounds
+RegionAxisBounds = List[float]
```
These are the accepted region axis bounds values (`lat_bounds` and `lon_bounds`). I think we should limit the scope of supported types to just a list of floats (either 1 value or 2 values [lower, upper]?)

tomvothecoder replied:
Fixed by limiting to a tuple of size two with validation function `_check_regional_bounds_input()`.

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -20,275 +23,311 @@
-        list of longitude values to validate
+    region_bounds : RegionAxisBounds
+        Region longitude bounds.
+    domain_bounds : xr.DataArray
```
For consistency, I decided to call everything domain (from the variable) and region (user-specified).

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -20,275 +23,311 @@
+    """
+    region_in_domain = _is_region_lon_in_domain(region_bounds, domain_bounds)
+    if not region_in_domain:
+        # TODO: Instead of the this implicit swapping of region axis,
+        # should we instead force the user to specify the correct lon bounds
+        # based on the axis orientation?
```
# TODO: Instead of the this implicit swapping of region axis,
# should we instead force the user to specify the correct lon bounds
# based on the axis orientation?

tomvothecoder commented on xcdat/variable.py:
```diff
@@ -63,3 +68,55 @@ def open_variable(dataset: xr.Dataset, name: str) -> xr.DataArray:
     return data_var
+
+
+def swap_lon_axis(var: xr.DataArray, to: Literal["180", "360"]):
```
This is the generic longitude swapping function I wrote for DataArrays

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -397,62 +428,55 @@ def spatial_average(xvariable,
-                                 ' the longitude axis domain.')
+    # TODO: Is `lat_bounds` and `lon_bounds` only two floating points for the
+    # upper and lower bounds or None? Or can a <2 or >2 values be specified?
+    if type(axis) == str:
+        axis = [axis]  # type:ignore
+    for axes in axis:
+        if axes not in SUPPORTED_AXES:
+            raise ValueError(
+                "Incorrect `axis` argument. Supported axis include: "
+                f"{', '.join(SUPPORTED_AXES)}."
+            )
```
This converts the `axis` str input to a list and checks if it is supported or not (currently just `lat`/`lon`).

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -397,62 +428,55 @@ def spatial_average(xvariable,
+                f"{', '.join(SUPPORTED_AXES)}."
+            )
+
+    # TODO: Do we need to perform validation on region lat bounds too?
+    if lon_bounds is not None:
+        domain_lon_bounds = xvariable.bounds.get_bounds("lon")
+        lon_bounds = _validate_region_lon_bounds(lon_bounds, domain_lon_bounds)
```
TODO: Do we need to perform validation on region lat bounds too?

pochedls replied:
I don't think this is necessary. A longitude value of -90 and 270 are technically equivalent, but the spatial averaging calculation depends on the longitude convention used in the dataset. This bit of code is intended to ensure our code works with either convention. Latitude doesn't have this problem: it is always -90 to 90. We could validate that user-specified bounds are in this range, but I don't think that is necessary.

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -397,62 +428,55 @@ def spatial_average(xvariable,
+        domain_lon_bounds = xvariable.bounds.get_bounds("lon")
+        lon_bounds = _validate_region_lon_bounds(lon_bounds, domain_lon_bounds)
+
+    # TODO: is_weighted function param is bool. Function doesn't support dataarray yet.
+    # TODO: if weights can be provided, it must align with is_weighted
```
TODO: is_weighted function param is bool. Function doesn't support dataarray yet.
TODO: if weights can be provided, it must align with is_weighted

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -353,11 +377,13 @@ def _averager(xvariable, weights, axis):
-                    weight='weighted'):
+def spatial_average(
+    xvariable: xr.DataArray,
+    axis: Union[List[SupportedAxes], SupportedAxes],
```
`axis` must be supplied

tomvothecoder replied:
> Is this in order to validate that the axes are in the list of SupportedAxes?

I think it is better to be explicit about which `axis` to average over. There is also defensive code to handle if the `axis` argument is incorrect.

> I noticed if I pass in a non-supported axis (e.g., `axis=['x', 'y']), I don't get an error. What is this doing if not checking the axis values passed in?

Not sure why you don't receive the `ValueError` because I do:
```
ValueError: Incorrect `axis` argument. Supported axis include: lat, lon.
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/Documents/Repositories/tomvothecoder/xcdat/qa/qa.py in 
      58 from xcdat.averager import spatial_average
      59 
----> 60 spatial_average(
      61     ts,
      62     axis=["x", "y"],

~/Documents/Repositories/tomvothecoder/xcdat/xcdat/averager.py in spatial_average(xvariable, axis, lat_bounds, lon_bounds, is_weighted)
    473     for axes in axis:
    474         if axes not in SUPPORTED_AXES:
--> 475             raise ValueError(
    476                 "Incorrect `axis` argument. Supported axis include: "
    477                 f"{', '.join(SUPPORTED_AXES)}."

ValueError: Incorrect `axis` argument. Supported axis include: lat, lon.
```

> We should update this to allow the user to specify a weighting array.

Got it, I will add this as a function param and update the algorithm accordingly.

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -353,11 +377,13 @@ def _averager(xvariable, weights, axis):
+    axis: Union[List[SupportedAxes], SupportedAxes],
+    lat_bounds: Optional[RegionAxisBounds] = None,
+    lon_bounds: Optional[RegionAxisBounds] = None,
+    is_weighted: bool = True,
```
`is_weighted` boolean instead of `weighted` as a str or dataarray (this wasn't supported yet).

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -299,52 +338,37 @@ def _averager(xvariable, weights, axis):
-    for i, dim in enumerate(weights.dims):
-        if len(weights[dim]) != len(xvariable[dim]):
-            ntile.append(len(xvariable[dim]))
-        else:
-            ntile.append(1)
-    weights = np.tile(weights, ntile)
-    # zero out missing data
-    weights[masked] = 0.
+    weights = weights.fillna(0)
+
+    # TODO: Below chunk of code should be extracted into a seperate function.
     # get averaging axis indices
```
I was able to refactor this entire chunk of code which is supposed to replicate `genutil.grower` using the initial `expand_dims` call. The `expand_dims` `dim` arg allows you to specify the dimension and size of the dimension to grow the dataarray.

https://xarray.pydata.org/en/stable/generated/xarray.DataArray.expand_dims.html

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -299,52 +338,37 @@ def _averager(xvariable, weights, axis):
         if dim not in weights.dims:
-            weights = weights.expand_dims(dim)
-    # tranpose to ensure axis ordering is the same
+            weights = weights.expand_dims({dim: size})
```
The specific line that takes care of tiling the dimensions for array expansion.

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -299,52 +338,37 @@ def _averager(xvariable, weights, axis):
+    # TODO: Below chunk of code should be extracted into a seperate function.
     # get averaging axis indices
     aaxis = []
     for i, dim in enumerate(xvariable.dims):
         if dim in axis:
             aaxis.append(i)
-    aaxis = tuple(aaxis)
+    aaxis = tuple(aaxis)  # type: ignore
     # weighted average over appropriate axes
     num = np.sum(weights * xvariable, axis=aaxis)
     den = np.sum(weights, axis=aaxis)
```
This chunk of code can be encapsulated in a separate function

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -1,9 +1,12 @@
 import numpy as np
+import xarray as xr
 
+# TODO: Reorder functions by callstack hierarchy
```
I plan on reordering the functions by callstack hierarchy.

It will be:
1. `spatial_averaging`
2. `_validate_region_lon_bounds()`
   - `_is_region_lon_in_domain`
   - `_swap_region_lon_axis()`
3. `_get_axis_weights()`
   - `_get_area_weights()` 
   - `_get_uniform_weights()`
   - `_scale_dim_to_region()`
4. ` _combine_axis_weights()`
5. `_averager()`

tomvothecoder replied:
@pochedls I would start with this comment to get a sense of how things are logically grouped.

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -20,275 +23,311 @@
+    return in_domain
+
+
+def _get_axis_weights(
```
This function has been refactored into separate smaller functions `_get_area_weights()` and `_get_uniform_weights()` + `_scale_dim_to_region()`.

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -20,275 +23,311 @@
+    xvariable: xr.DataArray,
+    lat_bounds: Optional[RegionAxisBounds],
+    lon_bounds: Optional[RegionAxisBounds],
+):
```
TODO: Add docstring here

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -20,275 +23,311 @@
+    weights: AxisWeights = {}
+
+    for dim_name, bounds in dims_bounds.items():
+        if bounds is not None and dim_name in SUPPORTED_AXES:
```
TODO: swap conditionals around so they it is more readable and can exit more quickly if needed

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -20,275 +23,311 @@
+        lon_bounds=[-170, -120]), by default None
+    is_weighted : bool, True
+        Weighting option to calculate weights proportional to area
+        ("weighted"), uniform weights ("unweighted").
```
TODO: change ("weighted") to True and ("unweighted") to False

pochedls commented on xcdat/averager.py:
```diff
@@ -0,0 +1,501 @@
+        Scaled dimension bounds based on regional selection.
+    """
+    # Adjustments for above / right of region
+    dimension[dimension[:, 0] > np.max(region_bounds), 0] = np.max(region_bounds)
```
I think this would fix most of the problem I mention above, but maybe not entirely.

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -0,0 +1,501 @@
+    weights = axis_weights[axis[0]]
+    # FIXME: By setting the initial value for weights to the first element (axes),
+    # it will multiple with itself in the loop below. Is this the intended behavior?
+    for axes in axis:
```
I've updated the function using this algorithm:
```
    weights = {axes: weights for axes, weights in axis_weights.items() if axes in axis}
    final_weights = reduce((lambda x, y: x * y), weights.values())
    return final_weights
```

It filters down the dictionary to the specified axes and then performs the multiplication using a `reduce()` so we don't need to set an initial value.

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -0,0 +1,501 @@
+
+    Notes
+    -----
+    This assumes a certain pathology (the axis orientation simply does
```
I can remove the note if it isn't relevant anymore

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -0,0 +1,543 @@
+
+    Notes
+    -----
+    This is a prototype and needs work to make it work more generically. It
+    currently only works with lat / lon. It might be useful to check the
+    latitude axis units.
+    """
+    BoundsByType = TypedDict(
+        "BoundsByType",
+        {"domain": Optional[xr.DataArray], "region": Optional[RegionAxisBounds]},
+    )
+    bounds: Dict[str, BoundsByType] = {
+        "lat": {"domain": xvariable.bounds.get_bounds("lat"), "region": lat_bounds},
+        "lon": {"domain": xvariable.bounds.get_bounds("lon"), "region": lon_bounds},
+    }
+    weights: AxisWeights = {}
+
+    for dim, bounds_by_type in bounds.items():
+        domain_bounds = bounds_by_type["domain"]
+        region_bounds = bounds_by_type["region"]
+
+        if domain_bounds is not None:
+            domain_bounds_copy = domain_bounds.copy()
+
+            if region_bounds is not None:
+                domain_bounds_copy = _scale_dim_to_region(
+                    domain_bounds_copy, region_bounds
+                )
+
+            if dim == "lat":
+                domain_bounds_copy = np.sin(np.radians(domain_bounds_copy))
+
+            # weights are generally the difference of the boundaries
+            # REVIEW: Is this always true for our purposes?
+            dim_wts = np.abs(domain_bounds_copy[:, 1] - domain_bounds_copy[:, 0])
+            dim_wts.attrs = domain_bounds_copy.attrs
+            dim_wts.name = dim + "_wts"
+            weights[dim] = dim_wts
+
+    return weights
```
I've updated the logic of `_get_area_weights()`.

It now stores the domain and region bounds in the `bounds` dict, instead of using `xvariable.bounds.bounds` (which gives a bounds dict for CF convention coordinate keys and axis).

I've statically defined the latitude key to "lat" and use `xvariable.bounds.get_bounds("lat"). You'll notice I check for `if dim == "lat"` now instead of `if dim in ["lat", "latitude"]`, since we loop over the `bounds` dict.

tomvothecoder replied:
I realized `xvariable.bounds.get_bounds("lat")` will throw and error if the bounds don't exist.
Instead, I will do:
```
domain_bounds = xvariable.bounds.bounds
bounds: Dict[str, BoundsByType] = {
    "lat": {"domain": bounds.get("lat"), "region": lat_bounds},
    "lon": {"domain": bounds.get("lon"), "region": lon_bounds}, 
}
```

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -0,0 +1,501 @@
+    return weights
+
+
+def _get_uniform_weights(xvariable: xr.DataArray) -> AxisWeights:
```
If spatial averaging is rarely done with uniform weights, we can eliminate it, along with the `is_weighted` param in `spatial_average()` (assume always weighted). Let me know if this sounds reasonable.

pochedls commented on xcdat/averager_accessor.py:
```diff
@@ -0,0 +1,506 @@
+                    # =========================================================
+                    if dim == "lon":
+                        # swap region + dim bounds to be 360-type
+                        domain_bounds[:] = self._swap_region_lon_axis(
```
@tomvothecoder - I think this line is problematic in some situations. Should `domain_bounds` be `domain_bounds_copy` (here and below)?

tomvothecoder replied:
Good catch, I just pushed a fix

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -0,0 +1,501 @@
+
+
+def _get_axis_weights(
+    xvariable: xr.DataArray,
```
Yup, I am going to refactor everything to `data_var`. I contemplated `var`, but that's too generic of a programming term.

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -0,0 +1,501 @@
+
+    Returns
+    -------
+    AxisWeights
```
These are Python type aliases, which are defined at the top of the module. `AxisWeights = Dict[Hashable, xr.DataArray]`

The reason I prefer type aliases is because it is a reusable. Type aliases are documented in the xCDAT docs.
We can also generate a tuple of options using the type alias if it is a `Literal`, which is great for input validation.

Example:
```python
#: List of supported axis for spatial averaging
SupportedAxes = Literal["lat", "lon"]
SUPPORTED_AXES: Tuple[SupportedAxes, ...] = get_args(SupportedAxes)

if axis not in SUPPORTED_AXIS:
   ....
```

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -0,0 +1,501 @@
+            if region_bounds is not None:
+                dim_bounds = _scale_dim_to_region(dim_bounds, region_bounds)
+
+            if dim_name in ["lat", "latitude"]:
```
This won't be a probably anymore, the for loop iterates over a dictionary with defined keys (only check "lat" now).

tomvothecoder commented on xcdat/averager.py:
```diff
@@ -0,0 +1,501 @@
+    return weights
+
+
+def _averager(
```
Hmmmm, this will be something good to take a closer look at. I'm not sure why CDAT performs poorly with larger domains. I'd have to dig into the algorithm to come up with an idea.

Also, it might help to use `timeit.repeat` to perform repeated runs to improve accuracy, unless that's what you already did. If we can leverage xarray `_xraverage()`, it would simplify our testing and implementation significantly!

pochedls commented on xcdat/averager_accessor.py:
```diff
@@ -0,0 +1,505 @@
+                        # specified (e.g., [0, 360] -> [360, 360]). In this case, make
+                        # the region bounds the upper/lower values of the dim bounds
+                        if region_bounds_arr[1] - region_bounds_arr[0] == 0:
+                            region_bounds_arr[1] = np.max(domain_bounds)
```
@tomvothecoder - I think this one needs to be updated to `domain_bounds_copy` too. I think this should work after that update.

tomvothecoder replied:
Thanks for the catch again, darn VSCode renaming lol.

tomvothecoder commented:
Hey @pochedls, I squashed the commits recently so your version of this branch may be really out of date. To simplify things, you can checkout `main`, delete this branch, and checkout the latest version of this branch.
 
**It would be great if you can review the `TODO:` comments in `spatial_avg.py`. I'm currently in the process of writing tests as well.**

Summary of Recent Changes:
- Updated the name of the decorator for the spatial averaging accessor from `ds.avg.spatial` to `ds.spatial.avg`
- Updated  the parent file name from `averager_accessor.py` to `spatial_avg.py`
- Added `xcdat.py`, which includes the `XCDATAccessor` to store all of our public accessor methods under a single namespace.
   - Spatial averaging can be accessed by `ds.spatial.avg` or `ds.xcdat.spatial_avg`
- On PR #105, `xcdat.open_dataset()` now limits datasets to a single variable

tomvothecoder commented:
TODOs from 9/29 meeting
- [ ] Finish up TODOs for code documentation
- [ ] Finish up QA of tests
- [ ] Make `axis` default to `["lat", "lon"]
- [ ] Time profiling for each of the individual methods -- a bit slower on variables with less data (add as issue after this merges)
- [ ] Figure out Dask error with parallelism
 
```python
NotImplementedError: xarray can't set arrays with multiple array indices to dask yet.
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
~/Documents/Repositories/tomvothecoder/xcdat/qa/PR98/refactor.py in <module>
      24 # ---------------------
      25 # NotImplementedError: xarray can't set arrays with multiple array indices to dask yet.
----> 26 avg = ds.spatial.avg("tas", axis, lat_bounds=lat_bounds, lon_bounds=lon_bounds)

~/Documents/Repositories/tomvothecoder/xcdat/xcdat/spatial_avg.py in avg(self, data_var_name, axis, weights, lat_bounds, lon_bounds)
    137             if lon_bounds is not None:
    138                 self._validate_region_bounds("lon", lon_bounds)
--> 139             weights = self._get_weights(axis, lat_bounds, lon_bounds)
    140 
    141         self._validate_weights(data_var, axis, weights)

~/Documents/Repositories/tomvothecoder/xcdat/xcdat/spatial_avg.py in _get_weights(self, axis, lat_bounds, lon_bounds)
    270                             domain_bounds, reg_bounds
    271                         )
--> 272                     dom_bounds = self._scale_domain_to_region(dom_bounds, reg_bounds)
    273 
    274                 if dim == "lat":

~/Documents/Repositories/tomvothecoder/xcdat/xcdat/spatial_avg.py in _scale_domain_to_region(self, domain_bounds, region_bounds)
    420             # Case 1: not wrapping around prime meridian.
    421             # Adjustments for above / right of region.
--> 422             d_bounds[d_bounds[:, 0] > r_bounds[1], 0] = r_bounds[1]
    423             d_bounds[d_bounds[:, 1] > r_bounds[1], 1] = r_bounds[1]
    424             # Adjustments for below / left of region.

/opt/miniconda3/envs/xcdat_dev/lib/python3.8/site-packages/xarray/core/dataarray.py in __setitem__(self, key, value)
    765                 for k, v in self._item_key_to_dict(key).items()
    766             }
--> 767             self.variable[key] = value
    768 
    769     def __delitem__(self, key: Any) -> None:

/opt/miniconda3/envs/xcdat_dev/lib/python3.8/site-packages/xarray/core/variable.py in __setitem__(self, key, value)
    852 
    853         indexable = as_indexable(self._data)
--> 854         indexable[index_tuple] = value
    855 
    856     @property

/opt/miniconda3/envs/xcdat_dev/lib/python3.8/site-packages/xarray/core/indexing.py in __setitem__(self, key, value)
   1244                 )
   1245                 if num_non_slices > 1:
-> 1246                     raise NotImplementedError(
   1247                         "xarray can't set arrays with multiple "
   1248                         "array indices to dask yet."

NotImplementedError: xarray can't set arrays with multiple array indices to dask yet.
  ```
